akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -179,15 +192,69 @@ private[akka] class Shard(
     None
   }
 
-  initialized()
+  val lease = settings.leaseSettings.map(
+    ls =>
+      LeaseProvider(context.system).getLease(
+        s"${context.system.name}-shard-$typeName-$shardId",
+        ls.leaseImplementation,
+        Cluster(context.system).selfAddress.hostPort))
 
-  def initialized(): Unit = context.parent ! ShardInitialized(shardId)
+  val leaseRetryInterval = settings.leaseSettings match {
+    case Some(l) => l.leaseRetryInterval
+    case None    => 5.seconds // not used
+  }
+
+  initialized(receiveCommand, ConstantFun.unitToUnit)
+
+  def initialized(next: Receive, onDone: () => Unit): Unit = {
+    lease match {
+      case Some(l) =>
+        tryGetLease(l)
+        context.become(awaitingLease(next, onDone))
+      case None =>
+        onDone()
+        context.parent ! ShardInitialized(shardId)
+        context.become(next)
+    }
+  }
+
+  def tryGetLease(l: Lease) = {
+    log.info("Acquiring lease {}", l.settings)
+    pipe(l.acquire(reason => self ! LeaseLost(reason)).map(r => LeaseAcquireResult(r, None)).recover {
+      case t => LeaseAcquireResult(acquired = false, Some(t))
+    }).to(self)
+  }
 
   def processChange[E <: StateChange](event: E)(handler: E => Unit): Unit =
     handler(event)
 
   def receive = receiveCommand
 
+  // Don't send back ShardInitialized so that messages are buffered in the ShardRegion
+  // while awaiting the lease
+  def awaitingLease(next: Receive, onDone: () => Unit): Receive = {
+    case LeaseAcquireResult(true, _) =>
+      log.info("Acquired lease. Ready to accept commands.")
+      onDone()
+      context.parent ! ShardInitialized(shardId)
+      context.become(next)
+    case LeaseAcquireResult(false, None) =>
+      log.error("Failed to get lease for shard type [{}] id [{}]. Retry in {}", typeName, shardId, leaseRetryInterval)
```

Comment:
`toString` of durations are not always pretty, `duration.toCoarsest` is one way, and we also have `PrettyPrintableDuration`

Reply:
:+1: